### PR TITLE
 (feat)Use nispor::NetState::retrieve_with_filter_async

### PR DESCRIPTION
**Pull request by jona42-ui**
As the first step to providing an async API in nmstate, this commit replaces usages of nispor::NetState::retrieve_with_filter() with nispor::NetState::retrieve_with_filter_async(). This change affects the files route.rs, route_rules.rs, show.rs, net_state.rs

This change improves the performance and responsiveness of the application by utilizing asynchronous retrieval of network state data.

Resolves:https://github.com/nmstate/nmstate/issues/2580

### CODE REVIEW

File: rust/src/lib/nispor/route.rs [hunk 1]
```diff
@@ -140,11 +149,6 @@ fn np_routetype_to_nmstate(np_route: &nispor::Route) -> RouteEntry {
             log::debug!("Got unsupported route {:?}", np_route);
         }
     }
-    // according to `man ip-route`, cwnd is useless without the lock flag, so
-    // we require both cwnd and its lock flag to consider cwnd as set.
-    let cwnd_lock = np_route.lock.unwrap_or(0) & (1 << RTAX_CWND) != 0;
-    route_entry.cwnd = if cwnd_lock { np_route.cwnd } else { None };
-
```
Comment (ffmancera) [inline rust/src/lib/nispor/route.rs]: Why are you dropping this?

File: rust/src/lib/nispor/route.rs [hunk 2]
```diff
@@ -196,10 +200,6 @@ fn np_route_to_nmstate(np_route: &nispor::Route) -> RouteEntry {
     route_entry.next_hop_addr = next_hop_addr;
     route_entry.metric = np_route.metric.map(i64::from);
     route_entry.table_id = Some(np_route.table);
-    // according to `man ip-route`, cwnd is useless without the lock flag, so
-    // we require both cwnd and its lock flag to consider cwnd as set.
-    let cwnd_lock = np_route.lock.unwrap_or(0) & (1 << RTAX_CWND) != 0;
-    route_entry.cwnd = if cwnd_lock { np_route.cwnd } else { None };
```
Comment (ffmancera) [inline rust/src/lib/nispor/route.rs]: An this?

**Review by ffmancera — CHANGES_REQUESTED** [review]

**Comment by jona42-ui** [comment]
my bad!
commiting them up stream now

**Comment by jona42-ui** [comment]
thanks @ffmancera

**Comment by ihuguet** [comment]
Not sure about what @cathay4t wanted exactly here, but I don't think we want to add a dependency to Tokio.
Also, I don't think that replacing public functions for async version of them is OK because it will break the users of those functions. Probably it's better to add the _async version of those functions, additionally to the synchronous ones.
Anyway, I'm a bit speculating, better to see what @cathay4t says.

**Comment by cathay4t** [comment]
Adding tokio dependency is fine as nispor already depend on it.

But please place `tokio` dependency to `query_apply` feature only in `lib/Cargo.toml`.

The code in general meet my expectation. Instead of change `pub fn retrieve()` to use `block_on`, try:

```rust
pub async fn retrieve_async() {
    // hold all old codes with awaits
}

pub fn retrieve() {
    block_on(Self::retrieve_async())
}
```

With this, we can change our `nm` to async also in the future without another `block_on()`.

**Comment by jona42-ui** [comment]
> Adding tokio dependency is fine as nispor already depend on it.
> 
> But please place `tokio` dependency to `query_apply` feature only in `lib/Cargo.toml`.
> 
> The code in general meet my expectation. Instead of change `pub fn retrieve()` to use `block_on`, try:
> 
> ```rust
> pub async fn retrieve_async() {
>     // hold all old codes with awaits
> }
> 
> pub fn retrieve() {
>     block_on(Self::retrieve_async())
> }
> ```
> 
> With this, we can change our `nm` to async also in the future without another `block_on()`.

@cathay4t  thanks so much
but I kindly have not understood the change requests.
I am sorry can you throw more light please.

**Comment by jona42-ui** [comment]
I placed the tokio dpendency in the `lib/Cargo.toml` do you mean I need to put it somewhere else?

**Comment by jona42-ui** [comment]
@cathay4t  it seems you reccomend  
tokio::runtime::Builder::new_multi_thread()
    .enable_all()
    .build()
    .unwrap()
    .block_on(your_async_func) here
https://developers.redhat.com/articles/2023/10/26/how-consume-and-expose-rust-async-crates#how_to_consume_rust_async_crates

to have more control of the tokio runtime

**Comment by jona42-ui** [comment]
also thanks for the writeup, it has helped me understand stuff well

**Comment by jona42-ui** [comment]
hi @cathay4t 
can you  take a look now?

**Comment by jona42-ui** [comment]
does this need unit tests? as rust unit fails

**Comment by jona42-ui** [comment]
hi @cathay4t 
made updates here.
waiting for a review
thanks

**Comment by jona42-ui** [comment]
although checks are failing, anything am missing?

**Comment by jona42-ui** [comment]
because it builds well

File: rust/src/lib/Cargo.toml [hunk 3]
```diff
@@ -61,6 +62,6 @@ serde_yaml = "0.9"
 
 [features]
 default = ["query_apply", "gen_conf", "gen_revert"]
-query_apply = ["dep:nispor", "dep:nix", "dep:zbus"]
+query_apply = ["dep:nispor", "dep:nix", "dep:zbus", "tokio"]
```
Comment (cathay4t) [inline rust/src/lib/Cargo.toml]: Please follow other example, use `dep:tokio`

File: rust/src/lib/nispor/route.rs [hunk 4]
```diff
@@ -41,23 +44,31 @@ pub(crate) fn get_routes(running_config_only: bool) -> Routes {
...
-        let mut rt_filter = nispor::NetStateRouteFilter::default();
+        let mut rt_filter = NetStateRouteFilter::default();
         rt_filter.protocol = Some(*protocol);
-        let mut filter = nispor::NetStateFilter::minimum();
+        let mut filter = NetStateFilter::minimum();
         filter.route = Some(rt_filter);
-        match nispor::NetState::retrieve_with_filter(&filter) {
-            Ok(np_state) => {
+
+        match task::spawn(async move {
```
Comment (cathay4t) [inline rust/src/lib/nispor/route.rs]: We do not need spawn here

**Review by cathay4t — CHANGES_REQUESTED** [review]

**Comment by cathay4t** [comment]
You find the CI log explaining the failure.

**Comment by jona42-ui** [comment]
@cathay4t  how best can i run the clib tests

**Comment by jona42-ui** [comment]
I have handled the rust-unit fail(not sure if I handled it well)

the rest seem strange to me especially the ...tier

**Comment by cathay4t** [comment]
This has been fixed by https://github.com/nmstate/nmstate/pull/2676

**Pull request closed without merging**